### PR TITLE
Function and test to determine the binary form of S and connectivities

### PR DESCRIPTION
**Pull request by laurentheirendt**
**I hereby confirm that I have:**

- [x] Tested my code on my own machine
- [x] Followed the guidelines in the [Contributing Guide](https://opencobra.github.io/cobratoolbox/docs/contributing.html)
- [x] Selected `develop` as a target branch (top left drop-down menu)

### CODE REVIEW

File: src/analysis/subspaces/determineBinaryMatrix.m [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+% Determine the binary form of the stoichiometric matrix S
+%
+% INPUT:
+%
+%    S:            Stoichiometric matrix of size m x n (m rows, n columns)
+%
+% OUTPUTS:
+%
+%    Shat:         Binary form of the stoichiometric matrix
+%    Shatabs:      Absolute value of Shat
```
Comment (tpfau) [inline src/analysis/subspaces/determineBinaryMatrix.m]: From the name of the function, I would actually expect it to only return Shatabs, or even just a logical sparse with `Sbin = S~=0`

File: src/analysis/subspaces/determineBinaryMatrix.m [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+% .. Author: - Laurent Heirendt, June 2018
+
+    Shat = S;  % initialize the binary matrix
+    Shat(S > 0) = 1;  % products, produced -> kin
+    Shat(S < 0) = -1;  % reactants, consumed -> kout
+
+    % determine the row connectivity
+    Shatabs = abs(Shat);
+    if nargout > 2
+        mconnect = sort(sum(Shatabs, 2), 'descend');
```
Comment (tpfau) [inline src/analysis/subspaces/determineBinaryMatrix.m]: sorting looks very odd here, I would expect the outcome to be linked to each respective row, while by sorting, you actually change that order.

File: src/analysis/subspaces/determineBinaryMatrix.m (hunk 1 above)
Comment (laurentheirendt) [reply src/analysis/subspaces/determineBinaryMatrix.m]: What other name do you suggest?

File: src/analysis/subspaces/determineBinaryMatrix.m (hunk 2 above)
Comment (laurentheirendt) [reply src/analysis/subspaces/determineBinaryMatrix.m]: I need them to be sorted, but I can specify that in the header.

Comment (tpfau) [reply src/analysis/subspaces/determineBinaryMatrix.m]: But why put that in this function? Couldn't you sort them afterwards?

File: src/analysis/subspaces/determineBinaryMatrix.m (hunk 1 above)
Comment (tpfau) [reply src/analysis/subspaces/determineBinaryMatrix.m]: `determineSignMatrix`?

File: src/analysis/subspaces/determineBinaryMatrix.m (hunk 2 above)
Comment (tpfau) [reply src/analysis/subspaces/determineBinaryMatrix.m]: And why are the other outputs not sorted? This is confusing (imo)

File: src/analysis/subspaces/determineBinaryMatrix.m [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+%    mconnect:     Compound connectivity (connectivity number; sum of the number of non-zero entries in a row)
+%    nconnect:     Reaction participation number (sum of the number of non-zero entries in a column)
+%    mconnectin:   Produced compound connectivity
+%    mconnectout:  Consumed compound connectivity
+%
+% .. Author: - Laurent Heirendt, June 2018
+
+    Shat = S;  % initialize the binary matrix
+    Shat(S > 0) = 1;  % products, produced -> kin
+    Shat(S < 0) = -1;  % reactants, consumed -> kout
```
Comment (tpfau) [inline src/analysis/subspaces/determineBinaryMatrix.m]: isnt this just:
Shat = sign(S) (which is about 50% faster)

File: src/analysis/subspaces/determineBinaryMatrix.m (hunk 1 above)
Comment (laurentheirendt) [reply src/analysis/subspaces/determineBinaryMatrix.m]: Ok I can change that.

File: src/analysis/subspaces/determineBinaryMatrix.m (hunk 2 above)
Comment (tpfau) [reply src/analysis/subspaces/determineBinaryMatrix.m]: I think a flag would make sense (but (imo) that flag should then sort everything, becuase currently this is somewhat of an inconsistency in the function.

Comment (laurentheirendt) [reply src/analysis/subspaces/determineBinaryMatrix.m]: Yes, I will implement such a flag.

**Comment by laurentheirendt** [comment]
Thanks for the review, @tpfau 👍

**Comment by rmtfleming** [comment]
Binary is only 0,1. With -1, 0, 1 it is a sign pattern matrix, or sign for short.

**Pull request merged**